### DISCUSSION TITLE
how are test developers meant to extend XUnit.Assert using partial class?

### DISCUSSION THREAD

**Discussion by mcallaghan-bsm** [xunit/xunit#2156] [Question] [answered]
Hi team and community,

I am deep in the rabbit hole with trying to write an "extension" of the Xunit.Assert capability in an internal project.

My goal:
* write a handful of local assertion extension capability for test utilities (they aren't super generic, so upstreaming is not sensible)
* ideally the callers don't know or care, and can simply run Assert... and find the methods

Namely, iterated over ALL of the options here: https://stackoverflow.com/questions/16194952/extending-xunit-assert-class-with-new-asserts

As I have learned, as xUnit evolved, various changes have come.  In xUnit 1, there were extension options through TestClass.AssertionsExtensions and other methods. https://xunit.net/docs/upgrade-extensions but this has been removed in xUnit 2.

A while ago I got the partial class method working, something simply like:
`CustomAssertions.cs`
``` c#
using ...

namespace Xunit
    public partial class Assert
    {
        public static async Task SomethingFunToCheckAsync() ...
    }
}
```

with the test project including
``` xml
        <PackageReference Include="xunit.assert.source" Version="2.4.1" />
```

As I understand it, partial class extensions are OK, only if they remain in the same assembly; But by definition here, the xunit source comes from a linked xunit.assert.source(.Asserts) assembly, and then we have our own local assembly...

The PROBLEM here, is that all sorts of analyzers complain about this pattern because we have TWO type definitions
```
FooTest.cs(312,19): warning CS0436: The type 'Assert' in '/root/.nuget/packages/xunit.assert.source/2.4.1/contentFiles/cs/netstandard1.1/Asserts/Assert.cs' conflicts with the imported type 'Assert' in 'xunit.assert, Version=2.4.1.0, Culture=neutral, PublicKeyToken=SNIP'. Using the type defined in '/root/.nuget/packages/xunit.assert.source/2.4.1/contentFiles/cs/netstandard1.1/Asserts/Assert.cs'. [/builds/dev/foo-projects/FooTest/FooTest.csproj]
```
https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0436

In addition, we see roslyn complaints in sonarqube of a similar nature
```
The type 'Assert' in '/root/.nuget/packages/xunit.assert.source/2.4.1/contentFiles/cs/netstandard1.1/Asserts/Assert.cs' conflicts with the imported type 'Assert' in 'xunit.assert, Version=2.4.1.0, Culture=neutral, PublicKeyToken=SNIP'. Using the type defined in '/root/.nuget/packages/xunit.assert.source/2.4.1/contentFiles/cs/netstandard1.1/Asserts/Assert.cs'.￼Why is this an issue?ROSLYN
> roslyn:CS0436Type conflicts with imported type
Code Smell
Major
No tags
ROSLYN
```

So to the point, let us consider: https://xunit.net/docs/nuget-packages
> "
  | xunit.assert.source | This package contains the xUnit.net assertion library (i.e., the Assert class) in source form. The Assert class is a partial class, which allows developers to write additional methods available directly from Assert.When you have multiple unit test libraries in your project, it is common practice to import this package into a "test utility" library where you write all your custom assertions, and then reference that "test utility" library from your unit test projects.
-- | -- | --
"

If we follow this guidance per the xunit docs, developers are left with this nightmare of roslyn/build errors/warnings CS0436.

I've started down the path of figuring out how to ignore this in the myriad of analyzers and build; I've figured out some ways to disable this particular message on an entire project or per member and that helps the `msbuild` side of things, however also having to juggle all the static analyzers and sonarqube scanning is getting extremely complicated ... but I've gone back to wondering ... _IS_ this the right way to extend Xunit assert?

Any advise is appreciated, thank you.!

**Comment by mcallaghan-bsm** [comment]
fwiw, I tried splitting out the Xunit.Assert partial class of our own into a separate project, and things got much worse, code can no longer distinguish properly
```
Ambiguous reference:
Xunit.Assert
Xunit.Assert
match (class) Xunit.Assert
```

**Comment by bradwilson** [comment, marked as answer]
I think the conflict is because you're bringing in both the binary and source asserts, presumably because you're still referencing the `xunit` meta-package (which itself references `xunit.assert`, `xunit.analyzers`, and `xunit.core`).

Your package references should look like this:

```xml
<PackageReference Include="xunit.analyzers" Version="0.10.0" />
<PackageReference Include="xunit.assert.source" Version="2.4.1" />
<PackageReference Include="xunit.core" Version="2.4.1" />
```

**Reply by mcallaghan-bsm** [reply]
Hey Brad,

An intreguing idea; but unfortunately that didn't quite finish the cake;

To expand on what I have now: (after your suggestion if I understood it correctly)

`Test.Utils` project, 
``` XML
    <ItemGroup>
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
        <PackageReference Include="Moq" Version="4.13.1" />
<!--        <PackageReference Include="xunit" Version="2.4.1" />-->
        <PackageReference Include="xunit.core" Version="2.4.1" />
        <PackageReference Include="xunit.assert.source" Version="2.4.1" />
        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
        <PackageReference Include="coverlet.collector" Version="1.2.0" />
    </ItemGroup>
```
``` c#
using ...
using Xunit.Sdk;

// ReSharper disable once CheckNamespace
namespace Xunit
    public partial class Assert
    {
        public static async Task SomethingFunToCheckAsync() ...
    }
}
```

Then some other `Foo` project,
``` xml
    <ItemGroup>
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
        <PackageReference Include="Moq" Version="4.13.1" />
<!--        <PackageReference Include="xunit" Version="2.4.1" />-->
        <PackageReference Include="xunit.core" Version="2.4.1" />
        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
    </ItemGroup>
```

``` C#
using Xunit;
using Xunit.Abstractions;

Assert.SomethingFunToCheckAsync(...)

Assert.True(...)

```

same ambiguous ref for all `Assert` usage attempts (whether they are to the core Xunit or our custom partial class additions)
```
Ambiguous reference:
Xunit.Assert
Xunit.Assert
match (class) Xunit.Assert
```

```
  Foo.cs(82, 13): [CS0433] The type 'Assert' exists in both 'Test.Utils, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' and 'xunit.assert, Version=2.4.1.0, Culture=neutral, PublicKeyToken=SNIP'
```

--
I also tried:
1. the `Test.Utils` project itself needs EITHER `xunit` or `xunit.core` (as it itself has unit tests to run against the assert partial class additions...) - either way, there are no project errors whether we choose either
2. in `Foo` project, it is unclear to me if we should or should not be including `xunit` or `xunit.core` or neither, as it DOES have a dependency on `Test.Utils`
3. in `Foo` project if we exclude `using Xunit;` nothing works and we get:
![2020-09-26_221218](https://user-images.githubusercontent.com/48924626/94354063-6d370d80-0045-11eb-918f-147c71128744.png)
, strangely, after we've moved the partial class addition into its own test project, `using Test.Utils;` has no affect on the ability to call `Assert.`

**Reply by bradwilson** [reply]
Can you provide a repro project?

**Reply by mcallaghan-bsm** [reply]
apologies, -- due to the struggles, I abandoned the idea of doing this -- and likely won't get back to it -- 

Would suggest that this is a use case for the XUnit team that would be worthwhile figuring out, and documenting, but of course this is up to your discretion.

**Reply by bradwilson** [reply]
Going to mark this as answered. Please let me know if you have any further questions.

**Reply by mcallaghan-bsm** [reply]
I know I am not in a position to assist with this further - but I'd say it is not answered - the documentation/sample needs to be improved - as following whatever is there today (or at the time) didn't work for us - but perhaps others from the community can come back to this one day and provide insight or "me too" for the issues
